source/common/formatter/substitution_formatter.cc

```diff
@@ -1140,5 +1154,27 @@ ProtobufWkt::Value StartTimeFormatter::formatValue(
       format(request_headers, response_headers, response_trailers, stream_info, local_reply_body));
 }
 
+EnvFormatter::EnvFormatter(const std::string& key) : key_(key) {}
+
+absl::optional<std::string> EnvFormatter::format(const Http::RequestHeaderMap&,
+    const Http::ResponseHeaderMap&,
+    const Http::ResponseTrailerMap&,
+    const StreamInfo::StreamInfo&,
+    absl::string_view) const {
+      const char* val = std::getenv(key_.c_str());
```

Comment:
I think we should avoid calling this every time we format a log entry. Since, if you `strace` it, `std::getenv` calls several syscalls. We can cache the values at initialization. See how we do this with HOSTNAME.

Reply:
That makes sense for performance, but do we still need to take care of its changes in running envoy?

Reply:
Maybe you should build a cache. If you see a new ENV name and it does not exist in the cache, resolve it by doing a syscall. It will solve perf issue and will work when a new config is pushed to Envoy.

Reply:
But typically you don't do "dynamic reloading" for a running process. Each process gets a copy of the environment variables.

Reply:
@dio Generally I agree with caching strategy on static reloading phase. But I'm not sure not to consider "dynamic reloading". I think that to watch `/proc/:pid/environ` with inotify with caching is better strategy. Is it not make sense?

Reply:
@Shikugawa could you provide a use-case on why you need that behavior?

Reply:
@dio Some users might annotate logs with environment variables. If they'd like to change this annotation, they must shutdown envoy process and reload it. Can't we say that these situations never happen?

Reply:
Sorry, but practically I'm not recommending this (I'm happy to be proved wrong). If we really want to bring a super flexible way to annotate logs with arbitrary labels, I believe we should not stand on env vars shoulder.

Another perspective: Say that you can do that, what if you accidentally change an env var and it changes the information inside log entries (it will bring headache to the log sink I think)?

Reply:
@dio It is make sense not to use env variables as flexible annotation. If these approach to annotate logs to be acceptable is not recommended, we don't have any situation to track the change state of them on envoy process. Accidentally changed them might bring chaos to log sink. So we shouldn't watch them. Thanks.

Reply:
OK, I will take a caching strategy. But, we also need to mention in the docs about its behavior (caching). I didn't know and got confused that `%HOSTNAME%` is currently cached and isn't reloaded dynamically, which seems not to be mentioned in the current document https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#config-access-log-format. (Sorry if I was missing something)